src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        private LowLevelConcurrentStack<SafeHeapHandle> handleCache = new LowLevelConcurrentStack<SafeHeapHandle>();
```

Comment:
I didn't see any impact on the GC- I had to do 100 million parallel iterations to even get a GC to happen. Are you testing against a copy of the original desktop code?

I'm not trying to dismiss your feedback- I'm trying to understand it. My difficulty is in understanding how your reticence applies when the method is _completely_ new. Speculating on the unmeasured impact is a useful exercise, but shouldn't prevent setting the starting point I think. I'm shipping a completely new implementation that I made in two steps: (1) correctness then (2) initial optimization. This is my starting position- from here I take it incrementally based on measured impact.

I agree that we may find subtle bugs- I expect them actually. More so from the logic changes in GetFullPath than the StringBuffer implementation. That said, I'd be even more worried about subtle issues messing with StringBuilder interop code to try and address the same set of issues (or even a subset).